Use the Fetch API to download the l10n files

Snuffleupagus opened the pull request:
Given that the Fetch API is supported since Node.js 18 we should be able to use it when downloading l10n files, which allows us to simplify the code and to make it fully `async`.

nicolo-ribaudo commented on external/importL10n/locales.mjs:
```diff
@@ -28,41 +27,32 @@ function normalizeText(s) {
+
+  // Remove any leading/trailing white-space.
+  const langCodes = normalizeText(content.trim()).split("\n");
+  // Remove all locales that we don't want to download below.
+  for (const langCode of [DEFAULT_LOCALE, ...EXCLUDE_LANG_CODES]) {
+    const i = langCodes.indexOf(langCode);
+    if (i > -1) {
+      langCodes.splice(i, 1);
+    }
+  }
+  return langCodes;
```
```suggestion
  // Remove all locales that we don't want to download below.
  return langCodes.filter(langCode =>
  	langCode !== DEAULT_LOCALE && !EXCLUDE_LANG_CODES.includes(langCode)
  );
```

or even better, make `EXCLUDE_LANG_CODES` a `Set` and use `.has` instead of `includes`.

Using `.splice` needs to re-iterate `langCodes` every time to move back all its elements by one, while with this `langCodes` is only iterated once.

Snuffleupagus replied:
Sure, we can do instead; thank you!

---

*Just in case it's not obvious:* Note that this code isn't run automatically in any part of the build process.
Instead a contributor will run this *manually* every once in a while, in order to update the l10n files.

nicolo-ribaudo commented on external/importL10n/locales.mjs:
```diff
@@ -71,41 +61,27 @@ function downloadLanguageFiles(root, langCode) {
-    });
-  });
+  // Download the necessary files for this language.
+  for (const fileName of files) {
```
The old implementation was downloading files in parallel, while this one is downloading them sequentially.

I don't know how many files this function downloads so I don't know if it actually has any impact, but in case it does you can still do it in parallel while using async/await:

```js
return Promise.all(files.map(async fileName => {
  const outputPath = path.join(outputDir, fileName);
  const url = MOZ_CENTRAL_ROOT + langCode + MOZ_CENTRAL_PDFJS_DIR + fileName;

  const response = await fetch(url);
  if (!response.ok) {
    // Not all files exist for each language. Files without translations
    // have been removed (https://bugzilla.mozilla.org/show_bug.cgi?id=1443175).
    continue;
  }
  const content = await response.text();

  await fs.promises.writeFile(outputPath, normalizeText(content), "utf8"));
}));
```

Snuffleupagus replied:
> I don't know how many files this function downloads so I don't know if it actually has any impact,

Currently, and for quite some time, it's only a single file per locale; see https://github.com/mozilla/pdf.js/blob/72b8b29147c3a19ccacc9db111df4925efcbe1bf/external/importL10n/locales.mjs#L72-L73

Besides, given how this code is used, absolute maximum speed isn't really a concern hence I'd prefer keeping this simple.

Snuffleupagus committed:
Use the Fetch API to download the l10n files

Given that the Fetch API is supported since Node.js 18 we should be able to use it when downloading l10n files, which allows us to simplify the code and to make it fully `async`.

Snuffleupagus committed:
Update l10n files

calixteman reviewed (approved):
LGTM. Thank you.
